experimental/overhead_matching/kimera_spectacular_data_provider.cc

```diff
@@ -0,0 +1,186 @@
+#include "experimental/overhead_matching/kimera_spectacular_data_provider.hh"
+
+#include <algorithm>  // for max
+#include <fstream>
+#include <opencv2/calib3d/calib3d.hpp>
+#include <opencv2/core/core.hpp>
+#include <string>
+#include <vector>
+
+#include "kimera-vio/imu-frontend/ImuFrontend-definitions.h"
+#include "kimera-vio/logging/Logger.h"
+
+namespace robot::experimental::overhead_matching {
+
+VIO::Timestamp vio_time_from_robot_time(const time::RobotTimestamp& robot_time) {
+    /// VIO timestamp is int64_t nanoseconds. time::RobotTimestamp is std::nano
+    return robot_time.time_since_epoch().count();
+}
+
+/* -------------------------------------------------------------------------- */
+SpectacularDataProviderInterface::SpectacularDataProviderInterface(const std::string& dataset_path,
+                                                                   const int& initial_k,
+                                                                   const int& final_k,
+                                                                   const VIO::VioParams& vio_params)
+    : DataProviderInterface(),
+      vio_params_(vio_params),
+      dataset_path_(dataset_path),
+      current_k_(std::numeric_limits<VIO::FrameId>::max()),
+      initial_k_(initial_k),
+      final_k_(final_k),
+      spec_log_(dataset_path) {
+    ROBOT_CHECK(!dataset_path_.empty(), "Dataset path for SpectacularDataProvider is empty.");
+
+    // Start processing dataset from frame initial_k.
+    // Useful to skip a bunch of images at the beginning (imu calibration).
+    ROBOT_CHECK(initial_k_ >= 0);
+
+    // Finish processing dataset at frame final_k.
+    // Last frame to process (to avoid processing the entire dataset),
+    // skip last frames.
+
+    if (final_k_ > spec_log_.num_frames()) {
+        LOG(WARNING) << "Provided final frame for KimeraSpectacularDataProvider was above the "
+                        "number of frames. Reducing to match size of dataset";
+        final_k_ = spec_log_.num_frames();
+    }
+
+    ROBOT_CHECK(final_k_ > 0);
+
+    ROBOT_CHECK(final_k_ > initial_k_, "Value for final_k is smaller than value for initial_k",
+                final_k_, initial_k_);
+    current_k_ = initial_k_;
+}
+
+SpectacularDataProviderInterface::~SpectacularDataProviderInterface() {
+    LOG(INFO) << "SpectacularDataProviderInterface destructor called.";
+}
+
+bool SpectacularDataProviderInterface::spin() {
+    if (!is_imu_data_sent_) {
+        // First, send all the IMU data. The flag is to avoid sending it several
+        // times if we are running in sequential mode.
+        if (imu_single_callback_) {
+            send_imu_data();
+        } else {
+            LOG(ERROR) << "Imu callback not registered! Not sending IMU data.";
+        }
+        is_imu_data_sent_ = true;
+    }
+
+    // Spin.
+    ROBOT_CHECK(final_k_ > initial_k_);
```

Comment:
this check doesn't seem necessary. The code below seems to not require this assumption. spin_once will reasonably return false.